### DISCUSSION TITLE
Can we create multiple jetstream consumers for a single subject

### DISCUSSION THREAD

**Discussion by YashasAnand** [nats-io/nats.net#653] [Q&A] [answered]
I am having a subject 'X.Y.Z' and in my application, I want to have 5 consumers consuming messages from 'X.Y.Z' subject.
Currently i am using c# application, where if i publish 10k messages to jetstream stream which is configured as below, all my 5 consumers are consuming 10k messages(total 50k messages consumed). 

await _natsConnection!.ConnectAsync();
        _natsJSContext = new NatsJSContext(_natsConnection);
        var streamConfig = new StreamConfig(
            name: _settings.NatsSettings.NatsJsStreamName,            
            subjects: [$"{_settings.NatsSettings.NatsJsStreamName}.>"] 
        )
        {
            Discard = StreamConfigDiscard.Old, 
            MaxBytes = 10L * 1024 * 1024  
        };

Note: This is a single node set up.

I was checking jetstream workqueues, but i believe we have subject filtering if we need 5 consumers to consume from same subject. But in my usecase we dont have any subject filtering because its just one static subject name.

Can anyone suggest if this is possible or not?

**Comment by mtmk** [comment]
You can have a single consumer and use the same consumer from 5 different clients to share the load.

**Comment by YashasAnand** [comment]
@mtmk I believe what you mean by 5 different clients would mean like 5 different instances of my application or 5 different pods if hosted in k8s? Pls can you confirm what you mean by clients?

What i am asking is if in a single application, i can loop through say 5 times to create a consumer with a specific configuration, where if I publish 10k messages , then 10k messages are load-balanced across 5 consumers running inside single instace of application.

If this is possible, pls can you share the possible configurtion parameters?

**Reply by mtmk** [reply]
yes I meant 5 application instances. but it could also be five different connections in a single process. Create one consumer on the server and use the same consumer from 5 different connections. does that make sense?

**Reply by YashasAnand** [reply]
@mtmk Thank you for the response, but i am unclear how to do this "Create one consumer on the server and use the same consumer from 5 different connections."? Also which retention policy do i have to use for creating multiple connections? 

Also are'nt connections costly to maintain? just for one subject if i create multiple connections, in a real-world application having many subjects, this wont scale i presume?

**Reply by mtmk** [reply]
in NATS consumer is a server side entity. you can create that once and start retrieving messages using the same consumer from multiple applications/connections. multiple connections may help with throughput depending on the use case. you have to experiment with it see what's best for you.

**Reply by mtmk** [reply, marked as answer]
maybe this helps as well: https://nats-io.github.io/nats.net/documentation/jetstream/intro.html#streaming-temporal-decoupling-between-the-publishers-and-subscribers

**Reply by YashasAnand** [reply]
ok thank you very much @mtmk , much appreciated,  I will go over this.

**Reply by YashasAnand** [reply]
@mtmk , had one doubt, i am using workqueue based retention & i have 2 consumers with subject filter introduced.
I initially tried with single instance of the app, my publisher publishes to topic X.Y.Z_0 & X.Y.Z_1.

This gets consumed by my consumers exactly half, lets say i publish 10 msgs, then for a single instance/client, each consumer processes 5 messages each.

Lets assume now i run 2 instances of this same application. Now there are 2 connections to broker, so will the Nats broker load-balance 10 messages to 5 & 5? 

because i tried this & my instance 1 received 6 messages, whereas instance 2 received 4 messages. How is nats broker doing the load-balancing here?

**Comment by mtmk** [comment]
I'd say overall percentage would be close to equal distribution but I'm out of my depth here when it comes to server internals. please feel free to jump on to `#jetstream` channel on [slack.nats.io](https://slack.nats.io/) or the [nats-server GitHub discussions](https://github.com/nats-io/nats-server/discussions) for more detail questions.

**Reply by YashasAnand** [reply]
ok , thank you again for your time @mtmk .
